### ISSUE TITLE
Question Generation in Recurrent Form

### ISSUE THREAD

**Issue by soda-lsq** [p208p2002/Transformer-QG-on-SQuAD#1]
Thanks for your wonderful Transformer-based QG code. Could you give some advice on how to implement recurrent BERT in the paper?

**Comment by p208p2002** [comment]
#### A1
Since BERT won't take **the last hidden output** as input, we do not need the loop you mentioned in training phase, it should done in data preprocess (split data input below from):
```
[CLS] [MASK] -> who
[CLS] who [MASK] -> wrote
[CLS] who wrote [MASK] -> the
[CLS] who wrote the [MASK] -> books
[CLS] who wrote the books [MASK] -> ?
[CLS] who wrote the books ? [MASK] -> [SEP]
``` 
consider that from is a batch, emphasize again BERT no need to wait  **the last hidden output** as input.

But do needs in predict phase, due to the design of Recurrent BERT need **the last token output** as input. 

I also note that there may some problem in your loop for decoding from `[MASK]`, about using the `torch.argmax`, please see below
```python
import torch

# assume the fake_last_hidden from "how are [MASK]"
# and model vocabs = ['how','are','you']
vocabs = ['how','are','you']
fake_last_hidden = torch.LongTensor([[5,1,2],[3,17,5],[6,7,10]]).unsqueeze(0)

# the output (from model) shape should be [batch_size,seq_len,vocab_size]
print(fake_last_hidden.shape)

# do argmax on the position of [MASK]
# torch.argmax(fake_last_hidden) <- this is worng, you will get `4` which is index of `17` (from fake_last_hidden.flat())
predict_ids = torch.argmax(fake_last_hidden,dim=-1) # index of each seq's max value 0->5, 1->17, 2->10
print(predict_ids)

# decoding using [MASK]'s hidden 
predict_id = predict_ids[0][-1] # 0 means first data in batch, carefully while batch_size > 1
vocabs[predict_id]
``` 

#### A2
It's a very good question! I've been discussing with my classmate before. The conclusion is BERT will re-encode in each step, but GPT do won't, that means the previous token's logits or hidden always changing  for BERT, this is reasonable, because **B**ERT is design for **B**idirectional. GPT fixed previous's hidden so that we could training in parallel, because GPT won't cheating to look the future information.

This is the operation mechanism that mention before (Self-Attention and Masked Self-Attention), also I've a little experimental for proof:
https://gist.github.com/p208p2002/537727bd8567f05564dd6b17c5638d83

**Comment by soda-lsq** [comment]
Hi Philip, 

Thanks again for your detailed explanation. Sorry for the late reply for trying to figure out the problems.

For A1, given the loop is conducted in the data pre-processing stage, I could be aware of how the loop is processed. I also used to break down one sentence into several sentence segments for training as the project did, in a different generation task, but still found the same problem that the generated sentence in the decoding stage is hard to stop. I find your explanation that the phenomenon is due to the model being trained in token-level rather than sentence-level is very reasonable. 

For A2, I see that Recurrent Bert could contain richer information due to the Bidirectional Re-encode, which keeps more information but low efficiency compared with GPT. And thanks for your demo. 

I briefly browsed the papers that cited this article, it seems that the Recurrent Pretrained-model in token-level generation task does not have a huge impact since not so much work follows the architecture. However, Recurrent Pretrained-model in sentence-level has more application. In some cases, the input sentence exceeds 512, to encode the long sentence, a recurrent layer is added upon the pre-trained layer to let the pre-trained information flow across segments. 

For now, I have understood all the problems. Thanks so much. This is the first time that my question has been answered in such a sufficient and detailed way, I really appreciate your help.
